File: standalone/dzen2-hlwm/lua/pipehandler.lua
-- luaposix available in AUR
local posix = require "posix"

local common = require('.common')
local helper = require('.helper')
local output = require('.output')

local _M = {}

-- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- -----
-- pipe

function _M.handle_command_event(monitor, event)
    -- find out event origin
    column = common.split(event, "\t")
    origin = column[0]

    tag_cmds = {'tag_changed', 'tag_flags', 'tag_added', 'tag_removed'}
    title_cmds = {'window_title_changed', 'focus_changed'}

    if origin == 'reload' then
        os.execute('pkill dzen2')
    elseif origin == 'quit_panel' then
        os.exit()
    elseif common.has_value(tag_cmds, origin) then
        output.set_tag_value(monitor)
    elseif common.has_value(title_cmds, origin) then
        output.set_windowtitle(column[2])
    end
end

function _M.init_content(monitor, process)
    -- initialize statusbar before loop
    output.set_tag_value(monitor)
    output.set_windowtitle('')

    local text = output.get_statusbar_text(monitor)
    process:write(text .. "\n")
    process:flush()
end

function _M.walk_content(monitor, process)    
    -- start a pipe
    command_in = 'herbstclient --idle'
    local pipe_in  = assert(io.popen(command_in,  'r'))
    local text = ''
  
    -- wait for each event 
    for event in pipe_in:lines() do
        _M.handle_command_event(monitor, event)    
    
        text = output.get_statusbar_text(monitor)
        process:write(text .. "\n")
        process:flush()
    end -- for loop
   
    pipein:close()
end

function _M.run_dzen2(monitor, parameters) 
    local command_out  = 'dzen2 ' .. parameters
    local pipe_out = assert(io.popen(command_out, 'w'))
    
    _M.init_content(monitor, pipe_out)
    _M.walk_content(monitor, pipe_out) -- loop for each event
        
    pipe_out:close()
end

function _M.detach_dzen2(monitor, parameters)
    local pid = posix.fork()

    if pid == 0 then -- this is the child process
        _M.run_dzen2(monitor, parameters)
    else             -- this is the parent process
        -- nothing
    end
end

function _M.detach_transset()
    local pid = posix.fork()

    if pid == 0 then -- this is the child process
        common.sleep(1)
        
        -- you may use either xorg-transset or transset-df instead
        -- https://github.com/wilfunctionyr/transset-df 
        os.execute('transset .8 -n dzentop >/dev/null') 
    else             -- this is the parent process
        -- nothing
    end
end


-- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- -----
-- return

return _M

File: standalone/dzen2-hlwm/php/pipehandler.php
<?php # using PHP7

require_once(__DIR__.'/output.php');

# ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----
# pipe

function handle_command_event($monitor, $event)
{
    // find out event origin
    $column = explode("\t", $event);
    $origin = $column[0];

    switch($origin) {
    case 'reload':
        system('pkill dzen2');
        break;
    case 'quit_panel':
        exit(1);
    case 'tag_changed':
    case 'tag_flags':
    case 'tag_added':
    case 'tag_removed':
        set_tag_value($monitor);
        break;
    case 'window_title_changed':
    case 'focus_changed':
        set_windowtitle($column[2]);
        break;
    }
}

function init_content($monitor, $process)
{   
    // initialize statusbar before loop
    set_tag_value($monitor);
    set_windowtitle('');
        
    $text = get_statusbar_text($monitor);
    fwrite($process, $text."\n");
    flush();
}

function walk_content($monitor, $process)
{       
    // start a pipe
    $command_in = 'herbstclient --idle';
    $pipe_in  = popen($command_in,  'r'); // handle
    
    while(!feof($pipe_in)) {
        # read next event
        $event = fgets($pipe_in);
        handle_command_event($monitor, $event);
        
        $text = get_statusbar_text($monitor);
        fwrite($process, $text);
        flush();
    }
    
    pclose($pipe_in);
}

function run_dzen2($monitor, $parameters) 
{ 
    $command_out  = "dzen2 $parameters";
    $pipe_out = popen($command_out, 'w');

    init_content($monitor, $pipe_out);
    walk_content($monitor, $pipe_out); // loop for each event

    pclose($pipe_out);
}

function detach_dzen2($monitor, $parameters)
{ 
    $pid = pcntl_fork();
    
    switch($pid) {         
    case -1 : // fork errror         
        die('could not fork');
    case 0  : // we are the child
        run_dzen2($monitor, $parameters); 
        break;
    default : // we are the parent             
        return $pid;
    }    
}

function detach_transset() 
{ 
    $pid = pcntl_fork();
    if ($pid == 0) { 
        sleep(1);
        system('transset .8 -n dzentop >/dev/null');
    }
}
